lua_modules/ftp/ftpserver.lua

```diff
@@ -0,0 +1,172 @@
+-- a simple ftp server
+local file,net,pairs,print,string,table = file,net,pairs,print,string,table
+local ftp, ftp_data, ftp_srv
+do
+local createServer = function (user, pass)
+local data_fnc, data_sock = nil, nil
+ftp_data = net.createServer(net.TCP, 180)
+ftp_data:listen(20, function (s) if data_fnc then data_fnc(s) else data_sock = s end end)
+ftp_srv = net.createServer(net.TCP, 180)
+ftp_srv:listen(21, function(socket)
+  local s = 0
+  local cwd = "/"
+  local buf = ""
+  local t = 0
+  socket:on("receive", function(c, d)
+    local a = {}
+    for i in string.gmatch(d, "([^ \r\n]+)") do
+      table.insert(a,i)
+    end
+    local a1,a2 = unpack(a)
+    if a1 == nil or a1 == "" then return end
+    if s == 0 and a1 == "USER" then
+      if a2 ~= user then
+        return c:send("530 user not found\r\n")
+      end
+      s = 1
+      return c:send("331 OK. Password required\r\n")
+    end
+    if s == 1 and a1 == "PASS" then
+      if a2 ~= pass then
+        return c:send("530 Try again\r\n")
+      end
+      s = 2
+      return c:send("230 OK.\r\n")
+    end
+    if s ~= 2 then
+      return c:send("530 Not logged in, authorization required\r\n")
+    end
+    if a1 == "CDUP" then
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if a1 == "CWD" then
+      if a2 == "." then
+        return c:send("257 \""..cwd.."\" is your current directory\r\n")
+      end
+      cwd = a2
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
```

Comment:
Suggest to change to:
~~~lua
      if a2 == "/" then
        cwd = a2
        return c:send("250 OK. Current directory is "..cwd.."\r\n")
      end
      return c:send("550 Failed to change directory.\r\n")
~~~
Needed to support web-browsers to retrieve the index of the root and prevent a CWD into a file.

Reply:
"cwd" in file is not main problem